File: building-control-mixed/hamr/src/main/component/building_control_mixed/BuildingControl/Fan_i_tcp_fan.scala
// #Sireum

package building_control_mixed.BuildingControl

import org.sireum._
import building_control_mixed._

// This file will not be overwritten so is safe to edit
object Fan_i_tcp_fan {

  def initialise(api: Fan_i_Initialization_Api): Unit = {
    api.sendfanAck(FanAck.Ok)
  }

  def handlefanCmd(api: Fan_i_Operational_Api, value : BuildingControl.FanCmd.Type): Unit = {
    api.logInfo(s"received fanCmd $value")

    val ack = FanNative.fanCmdActuate(value)

    api.sendfanAck(ack)

    api.logInfo(s"Actuation result: ${ack}")
  }

  def activate(api: Fan_i_Operational_Api): Unit = { }

  def deactivate(api: Fan_i_Operational_Api): Unit = { }

  def finalise(api: Fan_i_Operational_Api): Unit = { }

  def recover(api: Fan_i_Operational_Api): Unit = { }
}

@ext object FanNative {
  def fanCmdActuate(cmd: FanCmd.Type): FanAck.Type = $
}

File: building-control-mixed/hamr/src/c/ext-c/Fan_i_tcp_fan/Fan_i_tcp_fan.c
#include <Fan_i_tcp_fan_api.h>
#include <Fan_i_tcp_fan.h>
#include <ext.h>

// This file will not be overwritten so is safe to edit

Unit building_control_mixed_BuildingControl_Fan_i_tcp_fan_initialise_(STACK_FRAME_ONLY) {
  DeclNewStackFrame(caller, "Fan_i_tcp_fan.c", "", "building_control_mixed_BuildingControl_Fan_i_tcp_fan_initialise_", 0);

  api_send_fanAck__building_control_mixed_BuildingControl_Fan_i_tcp_fan(SF 
    building_control_mixed_BuildingControl_FanAck_Type_Ok);
}

Unit building_control_mixed_BuildingControl_Fan_i_tcp_fan_handlefanCmd_(
  STACK_FRAME
  building_control_mixed_BuildingControl_FanCmd_Type value) {
  DeclNewStackFrame(caller, "Fan_i_tcp_fan.c", "", "building_control_mixed_BuildingControl_Fan_i_tcp_fan_handlefanCmd_", 0);

  String str = string("received fanCmd "); 
  building_control_mixed_BuildingControl_FanCmd_Type_string_(SF str, value);
  api_logInfo__building_control_mixed_BuildingControl_Fan_i_tcp_fan(SF
    str);

  // always return 'Ok'
  building_control_mixed_BuildingControl_FanAck_Type ack = 
    building_control_mixed_BuildingControl_FanAck_Type_Ok;
    
  api_send_fanAck__building_control_mixed_BuildingControl_Fan_i_tcp_fan(SF 
    ack);
  
  str = string("Actuation result: "); 
  building_control_mixed_BuildingControl_FanAck_Type_string_(SF str, ack);
}

Unit building_control_mixed_BuildingControl_Fan_i_tcp_fan_finalise_(STACK_FRAME_ONLY) {
  DeclNewStackFrame(caller, "Fan_i_tcp_fan.c", "", "building_control_mixed_BuildingControl_Fan_i_tcp_fan_finalise_", 0);
  
  // example finalise method
}